File: java/01_walking_skeleton/agent-infrastructure/src/main/java/com/workshop/agent/infrastructure/OllamaLanguageModel.java
package com.workshop.agent.infrastructure;

import com.workshop.agent.application.LanguageModel;
import com.workshop.agent.application.Message;

import java.util.List;

public class OllamaLanguageModel implements LanguageModel {
    public OllamaLanguageModel(String baseUrl, String model) {
    }

    @Override
    public Message prompt(List<Message> message) {
        return null;
    }
}

File: csharp/03_back_and_forth_chat/Agent/Agent.Infrastructure/OllamaLanguageModel.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using Agent.Application;

namespace Agent.Infrastructure;

public class OllamaLanguageModel(string serverUrl, string model) : ILanguageModel
{
    private readonly HttpClient _httpClient = new() { BaseAddress = new Uri(serverUrl) };

    private static readonly JsonSerializerOptions JsonSerializerOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
    };

    public Message Prompt(IEnumerable<Message> messages)
    {
        var httpResponse = _httpClient.PostAsync("/v1/chat/completions",
            JsonContent.Create(
                new ChatRequest(model, messages.ToList()),
                new MediaTypeHeaderValue("application/json"),
                JsonSerializerOptions
            )
        ).GetAwaiter().GetResult();
        
        httpResponse.EnsureSuccessStatusCode();
        
        var chatCompletion = JsonSerializer.Deserialize<ChatCompletionResponse>(
            httpResponse.Content.ReadAsStringAsync().GetAwaiter().GetResult(),
            JsonSerializerOptions
        );

        return chatCompletion?.Choices.FirstOrDefault()?.Message!;
    }
}